### ISSUE TITLE
Work on menu option: Delete a startup item

### ISSUE THREAD

**Issue by dEhiN** [dEhiN/Demord#5]
In the function json_editor, there is a menu option to delete a startup item. Currently, when selected, a message is printed out saying, "This functionality hasn't been implemented yet". This needs to be planned out in terms of how it will be done, including any helper functions that might be necessary, and how the existing startup data will be rewritten without the delete item.

**Comment by dEhiN** [comment]
Starting work on this issue

**Comment by dEhiN** [comment]
Copying commit notes from commit 4537f8a:

```
Updated function json_pruner to print out the
item requested to be deleted as well as the rest
of the items data. Also added a call to the
function generate_user_edited_data.

However, passing in just the Items array causes
the JSON data validation to fail. Will need to
figure out a solution. Possible options are:

1. Modify the generate_user_edited_data
function to allow for an Items array to be passed
in.

2. Modify the json_pruner function to pass in
updated JSON data.

3. Modify a mix of both functions
to allow for a single startup item to be passed in
but specified to be removed.

4. Create a new function to take the Items array
and make a fully formed startup JSON object from
it.

Most likely, will go with option 3 since the
generate_user_edited_data is meant to be used for
all scenarios where the user is editing existing
startup data. There is already the ability to
pass in just a single startup item and also to
specify if that startup item is to be added to the
existing full startup data or replace an existing
item. It shouldn't be hard to modify the function
to allow the startup item to be removed and
specify that.
```

**Comment by dEhiN** [comment]
Going with option 3 from the comment above:

- [x] Modify the function `generate_user_edited_data` so a calling function can pass in a single startup item and specify that it needs to be removed from the existing startup data
- [ ] If necessary, update the scenario validation functions `data_validation_scenario` and `match_scenario`
- [ ] Call the modified `generate_user_edited_data` function from `json_pruner` 
- [ ] Test that it works

The above are the steps to take to resolve this issue and close the branch `dEhiN/issue5`.

ETA: See comments below for updated versions of this list

**Comment by dEhiN** [comment]
Taken from commit f46289c7 for what still need to be done:

1. Determine if the scenario where `item_type` is _F_ is needed or can be removed from the following functions: `generate_user_edited_data` `data_validation_scenario`, and `match_scenario`
2. Update the docstrings accordingly in all three functions
3. Refactor the relevant code in all three functions
4. Code in the functionality in `generate_user_edited_data` to actually remove an item and return the updated JSON data

From that commit and the code changes made to the module `data_generate`, the following notes about the function `generate_user_edited_data` can be gathered:

- the function took, in its previous declaration, 2 dictionary parameters related to JSON data and 1 boolean parameter
- the first was called `modified_json_data` and could be either the full startup data or a single startup item
- the second was called `item_add` and was used to determine if a startup item was to completely replace an existing item in the startup item arguments list, or to be added to the end of the startup item arguments list
- the last was called `orig_json_data` and would would have the original startup data when the first parameter was a single startup item
- the function declaration no longer takes the boolean parameter, and, instead, takes a string parameter called `item_type`, which can only be the values _A_, _D_, _R_, or _F_
- the first value will be used to add a startup item to the existing list, the second to delete an existing startup item from the existing list, and the third to replace an existing startup item, and the fourth to state that the `modified_json_data` parameter consists of full startup data
- however, this last value and scenario may not be necessary
- with the old boolean parameter, based on the combination of all three parameters, one of three scenarios was determined corresponding to the values _A_, _R_, and _F_ for `item_type`
- when the scenario was equivalent to _F_, the function just returned a copy of the `modified_json_data` variable

This is why step 1 above is listed the way it is; I don't remember the reasoning behind created a scenario where fully-formed, startup JSON data is passed in. The two possibilities I can think of are:

1. The function `generate_user_edited_data` calls a function that validates both parameters `modified_json_data` and `orig_json_data` (if it exists) against the two JSON schemas written up for the full startup data and a single startup item
2. Due to the way data structures like dictionaries and lists work in Python, a true copy of the contents of the `modified_json_data` variable is created and returned

Most likely, both possibilites are the reasons for having fully-formed, startup JSON data passed in. The second possibility is accomplished by using the Python `deepcopy` function in the `copy` module, which will create a true copy of the data structure and all sub-data structures.

**Comment by dEhiN** [comment]
Amalgamating the logic and lists from both previous comments, I think I can safely say the next steps for this issue are:

1. Refactor the function `data_validation_scenario` to utilize the `item_type` variable
2. Refactor the function `data_validation_scenario` to screen for all four scenarios, each of the  four values of the `item_type` variable
3. Refactor the function `match_scenario` to account for the four scenarios
4. Decide if the function `generate_user_edited_data` will remove a startup item and update the startup JSON data, or if the calling function is responsible for that

While working on step 1, the ordering, or number assigned, to each scenario may need to change to reflect the if-block in `generate_user_edited_data`.

**Comment by dEhiN** [comment]
Adding in a step 5 to the above list:

5. Update all calls to `generate_user_edited_data` to pass in the correct parameters, mainly, that the `item_type` parameter is now a string

While working on steps 1-3, testing revealed that, for the existing scenarios, the calls to `generate_user_edited_data` causes an error because the calls are still based on when the second parameter was a boolean.

**Comment by dEhiN** [comment]
Need to plan out for `match_scenario` the different situations or scenarios. There are 4 scenarios and 5 error scenarios:

_Valid Scenarios:_
1. Need to add a single startup item
2. Need to remove a single startup item
3. Need to replace a single startup item that already exists
4. Need to update the full JSON data

_Error Scenarios:_
1. Expected original JSON data but that cannot be found
2. Original JSON data passed in is not properly formed
3. Expected modified JSON data to be a single startup item but didn't receive that
4. Expected modified JSON data to be a full startup data but didn't receive that
5. The modified JSON data passed in is not properly formed

The 5 keys in the dictionary `data_validation` used by `match_scenario` to determine each scenario are: `Item-Type`, `Orig-Exists`, `Orig-Valid`, `Mod-Single`, and `Mod-Valid`

Their descriptions are as follows:
1. `Item-Type` -> Used to determine whether to add/delete/replace a single startup item or replace the full startup data
2. `Orig-Exists` -> For the scenarios where a single startup item is being worked on, was the original JSON data given or not
3. `Orig-Valid` -> If the original JSON data was given, is it valid
4. `Mod-Single` -> Does the parameter with the new JSON data have a single startup item or full JSON data
5. `Mod-Valid` -> Is the new JSON data valid or not

Although key 1 can be used to determine which type of new JSON data was passed in, key 4 needs to also be checked in case key 1 and what was actually given as new JSON data don't match. The same is the case with key 5; depending on what key 4 contains, the new JSON data will be validated against the JSON schemas for either a single startup item or the full startup data.

Accepted values for each key are:
1. `Item-Type`: A, D, R, F
2. `Orig-Exists`: True, False
3. `Orig-Valid`: True, False*
4. `Mod-Single`: True, False
5. `Mod-Valid`: True, False

*If key 2 if False, then key 3 will also be False. This is because keys 2 to 5 are set to False as defaults. 

Amalgamating all the above information together, each valid and error scenario will constitute the following combination of key values. Note, a value of '??' means the key doesn't need to be checked for that scenario:

_VS_1_:
`Item-Type`: A
`Orig-Exists`: T
`Orig-Valid`: T
`Mod-Single`: T
`Mod-Valid`: T

_VS_2_:
`Item-Type`: D
`Orig-Exists`: T
`Orig-Valid`: T
`Mod-Single`: T
`Mod-Valid`:  T

_VS_3_:
`Item-Type`: R
`Orig-Exists`: T
`Orig-Valid`: T
`Mod-Single`: T
`Mod-Valid`:  T

_VS_4_:
`Item-Type`: F
`Orig-Exists`: F
`Orig-Valid`: F
`Mod-Single`: F
`Mod-Valid`: T

_ES_1_:
`Item-Type`: A/D/R
`Orig-Exists`: F
`Orig-Valid`: ??
`Mod-Single`: ??
`Mod-Valid`: ??

_ES_2_:
`Item-Type`: A/D/R
`Orig-Exists`: T
`Orig-Valid`: F
`Mod-Single`: ??
`Mod-Valid`: ??

_ES_3_:
`Item-Type`: A/D/R
`Orig-Exists`: ??
`Orig-Valid`: ??
`Mod-Single`: F
`Mod-Valid`: ??

_ES_4_:
`Item-Type`: F
`Orig-Exists`: ??
`Orig-Valid`: ??
`Mod-Single`: T
`Mod-Valid`:  ??

_ES_5_:
`Item-Type`: ??
`Orig-Exists`: ??
`Orig-Valid`: ??
`Mod-Single`: ??
`Mod-Valid`: F

**Comment by dEhiN** [comment]
Further thoughts to the comment above:
- Rework data_validation dictionary to only have the 4 Boolean keys
- Have match_scenario take in the dictionary and the item_type string variable as two parameters
- Can then create decision-branching based on a mix of the dictionary and string parameters

Rewriting the 5 error scenarios in a condensed format, we get:
- ES_1: Item-Type: A;D;R / Orig-Exists: F / Orig-Valid: ?? / Mod-Single: ?? / Mod-Valid: ??
- ES_2: Item-Type: A;D;R / Orig-Exists: T / Orig-Valid: F / Mod-Single: ?? / Mod-Valid: ??
- ES_3: Item-Type: A;D;R / Orig-Exists: ?? / Orig-Valid: ?? / Mod-Single: F / Mod-Valid: ??
- ES_4: Item-Type: F / Orig-Exists: ?? / Orig-Valid: ?? / Mod-Single: T / Mod-Valid:  ??
- ES_5: Item-Type: ?? / Orig-Exists: ?? / Orig-Valid: ?? / Mod-Single: ?? / Mod-Valid: F

Rewriting them for the minimum needed, we get:
- ES_5: Mod-Valid = False
- ES_4: Mod-Single = True && Item-Type = "F"
- ES_3: Mod-Single = False && (Item-Type = "A" || Item-Type = "D" || Item-Type = "R")*
- ES_1: Orig-Exists = False && (Item-Type = "A" || Item-Type = "D" || Item-Type = "R")*
- ES_2: Orig-Exists = True && Orig-Valid = False*

*ES_3 / ES_1: Another way to write the bracketed conditions is to say Item-Type <> "F" since Item-Type will already have been validated to only contain A/D/R/F before being passed in
*ES_2: It's not actually necessary to check Item-Type for error scenario 2 since the scenario is where the original JSON data passed in is invalid

Summarizing that, we can write the pseudo-code as the following. After each error scenario, a return will probably be needed to break out of the decision branching:
```
if Mod-Valid = False: Error Scenario 5
else:
-> if Item-Type = "F":
-> -> if Mod-Single = True: Error Scenario 4
-> else:
-> -> if Mod-Single = False: Error Scenario 3
-> -> if Orig-Exists = False: Error Scenario 1
-> -> else:
-> -> -> if Orig-Valid = False: Error Scenario 2

```

**Comment by dEhiN** [comment]
Rewriting the 4 valid scenarios in a condensed format, we get:
- VS_1: Item-Type: A / Orig-Exists: T / Orig-Valid: T / Mod-Single: T / Mod-Valid: T
- VS_2: Item-Type: D / Orig-Exists: T / Orig-Valid: T / Mod-Single: T / Mod-Valid:  T
- VS_3: Item-Type: R / Orig-Exists: T / Orig-Valid: T / Mod-Single: T / Mod-Valid:  T
- VS_4: Item-Type: F / Orig-Exists: F / Orig-Valid: F / Mod-Single: F / Mod-Valid: T

Rewriting them for the minimum needed, we get:
- VS_1: (Orig-Exists = True && Orig-Valid = True && Mod-Single = True && Mod-Valid = True) && Item-Type = "A"
- VS_2: (Orig-Exists = True && Orig-Valid = True && Mod-Single = True && Mod-Valid = True) && Item-Type = "D"
- VS_3: (Orig-Exists = True && Orig-Valid = True && Mod-Single = True && Mod-Valid = True) && Item-Type = "R"
- VS_4: (Mod-Single = False && Mod-Valid = True) && Item-Type = "F"*

*Since the Item-Type needs to be F and the modified JSON data needs to be valid and not a single startup item, that's all we need to check for valid scenario 4

Summarizing that, we can write the pseudo-code as the following. After each error scenario, a return will probably be needed to break out of the decision branching
```
if Mod-Valid = True:
-> if Mod-Single = False:
-> -> if Item-Type = "F": Valid Scenario 4
-> else:
-> -> if Orig-Exists = True:
-> -> -> if Orig-Valid = True:
-> -> -> -> if Item-Type = "A": Valid Scenario 1
-> -> -> -> if Item-Type = "D": Valid Scenario 2
-> -> -> -> if Item-Type = "R": Valid Scenario 3
```

**Comment by dEhiN** [comment]
Amalgamating both pseudo-code decision branching blocks from the last two comments, we get:
```
if Mod-Valid = False: 
    Error Scenario 5
else Mod-Valid = True:
    if Item-Type = "F":
        if Mod-Single = False: 
             Valid Scenario 4
        else Mod-Single = True:
            Error Scenario 4
    else Item-Type <> "F":
        if Mod-Single = True:
            if Orig-Exists = True:
                if Orig-Valid = True:
                    if Item-Type = "A": 
                        Valid Scenario 1
                    if Item-Type = "D": 
                        Valid Scenario 2
                    if Item-Type = "R": 
                        Valid Scenario 3
                else Orig-Valid = False: 
                    Error Scenario 2
            else Orig-Exists = False: 
                Error Scenario 1
        else Mod-Single = False: 
            Error Scenario 3

```

Cleaning that up, we get:
```
if Mod-Valid = False: 
    Error Scenario 5
else:
    if Item-Type = "F":
        if Mod-Single = False: 
             Valid Scenario 4
        else:
            Error Scenario 4
    else:
        if Mod-Single = True:
            if Orig-Exists = True:
                if Orig-Valid = True:
                    if Item-Type = "A": 
                        Valid Scenario 1
                    else if Item-Type = "D": 
                        Valid Scenario 2
                    else: 
                        Valid Scenario 3
                else: 
                    Error Scenario 2
            else: 
                Error Scenario 1
        else: 
            Error Scenario 3
```

Revisiting the thoughts from before:
- Keep the data_validation dictionary as it is
- Have match_scenario expand the dictionary to 4 variables to maybe make it easier to read the branching block

**Comment by dEhiN** [comment]
Fixed `match_scenario` through commit 8690bae6 but discovered another issue. While testing, noted that the option to save a modified startup item to disk doesn't work properly. It saves the data but as a new item. The good news is that the Total Items JSON property is updated correctly. However, the option to save the startup data to disk does not work correctly.

Specifically, the menu options are as follows:

1. Main menu > 4: Edit the existing startup file > 6: Save full startup data to disk
2. Main menu > 4: Edit the existing startup file > 1: Edit startup item 1 > 6: Save startup item data to disk

In both situations, the data needs to be modified. Additionally, any startup item could be edited. For example, with the following steps:

1. Select to edit startup item 1
2. Change the name of the startup item
3. Select to save the startup item data to disk

The modified startup item 1 will be saved but as a new startup item and the Total Items property updated. So, if using the default startup data which has 3 items, the above steps would create a 4th item. This fourth item will be item 1 with the changed name.

With the following steps, the results are what is expected:

1. Select to edit startup item 1
2. Change the name of the startup item
3. Select to return to the previous menu
4. Select to save the startup data to disk

In this case, the full startup data will be saved with the correctly modified item. So, if using the default startup data which has 3 items, 3 items will be saved and item 1 will have the updated name.

**Comment by dEhiN** [comment]
The above issue is fixed - see commit 0c57a17b.

**Comment by dEhiN** [comment]
This issue is now resolved completely - see commits b0e1ffeb to bce91cec. Noted an edge case in commit 68216ec4 which basically involves when a user makes a change but doesn't save right away. Will create a new issue for this.
